packages/api/src/api/models/Rules/Rules.ts

```diff
@@ -0,0 +1,114 @@
+      dimension: databaseRule.dimension,
+      threshold: databaseRule.threshold,
+      destinationTimeSeries: databaseRule.destination_time_series,
+      updatedAt: databaseRule.updated_at.getTime(),
+      createdAt: databaseRule.created_at.getTime(),
+      enabled: databaseRule.enabled,
+    };
+  };
+
+  public insertRules = async (
+    rules: Rule[],
+    rulesetId: number,
+    projectId: number,
+    parentSpan: Span
+  ) => {
+    return await Promise.all(
+      rules.map(
+        async (rl: RuleData): Promise<Rule> => {
+          const { rows } = await PostgresDatasource.query<RuleDBSchema>(
+            {
+              name: 'insert-rules',
+              text: `INSERT INTO rules 
+              (rule_set_id, dimension, threshold, enabled)
+            SELECT 
+              rule_sets.rule_set_id, $2, $3, $4
+            FROM rule_sets
+            WHERE rule_sets.rule_set_id = $1 AND rule_sets.monitor_id = (
+              SELECT monitor_id 
+              FROM monitors 
+              WHERE monitors.monitor_id = rule_sets.monitor_id AND monitors.project_id = $6)
+            RETURNING 
+              rule_id, rule_set_id, dimension, threshold, destination_time_series, created_at, updated_at, enabled`,
+              values: [
+                rulesetId,
+                rl.dimension,
+                rl.threshold,
+                rl.enabled,
+                projectId,
+              ],
+            },
+            parentSpan
+          );
+          const addedRule: RuleDBSchema = rows[0];
+          return this.mapDatabaseRule(addedRule);
+        }
+      )
+    );
+  };
+
+  public updateRules = async (
+    rules: Rule[],
+    rulesetId: number,
+    projectId: number,
+    parentSpan: Span
+  ) => {
+    const queries: QueryConfig[] = [
+      {
+        name: 'clean-rules',
+        text: `DELETE FROM rules 
+      WHERE rule_set_id = $1;`,
+        values: [rulesetId],
+      },
+    ];
+    rules.forEach((rl: RuleData) => {
+      queries.push({
+        name: 'insert-rules',
+        text: `INSERT INTO rules 
+              (rule_set_id, dimension, threshold, destination_time_series, enabled)
+            SELECT 
+              rule_sets.rule_set_id, $2, $3, $4, $5
+            FROM rule_sets
+            WHERE rule_sets.rule_set_id = $1 AND rule_sets.monitor_id = (
+              SELECT monitor_id 
+              FROM monitors 
+              WHERE monitors.monitor_id = rule_sets.monitor_id AND monitors.project_id = $6)
+            RETURNING 
+              rule_id, rule_set_id, dimension, threshold, destination_time_series, created_at, updated_at, enabled`,
+        values: [
+          rulesetId,
+          rl.dimension,
+          rl.threshold,
+          0,
+          rl.enabled,
+          projectId,
+        ],
+      });
+    });
+    const result = await PostgresDatasource.runTransactionalQueries(
+      queries,
+      parentSpan
+    );
+    const ruleResults: Rule[] = [];
+    result.forEach(qr => {
```

Comment:
Can we rewrite the forEach to be a map, making it functional?

Reply:
because we dont know what each query might return, making it functional generates typescript errors in terms of it might return void or Rule, so i think we should keep it this unless you have a suggestion on how to avoid all the typing errors? [](url)
```suggestion
Type 'void[]' is not assignable to type 'Rule[]'.
  Type 'void' is not assignable to type 'Rule'.
```

Reply:
Aha. But you're silently doing a filter in here ;) 

So you would do 
```
result.filter(qr => qr).map(mapDatabaseRule);
```

Reply:
fixed